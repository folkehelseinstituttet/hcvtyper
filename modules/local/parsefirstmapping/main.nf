process PARSEFIRSTMAPPING {
    tag "$meta.id"
    label 'process_single'

    // Environment with R tidyverse and seqinr packages from the conda-forge channel. Created using seqera containers.
    // URLs:
    // Docker image: https://wave.seqera.io/view/builds/bd-5358395134867368_1?_gl=1*1vaclhd*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    // Singularity image: https://wave.seqera.io/view/builds/bd-0225dab2b8112adf_1?_gl=1*111m8r6*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/b5/b5d7b977f4b94903794f65bbd852248eef87ce608c5eef34d605afabf514f397/data':
        'community.wave.seqera.io/library/r-seqinr_r-tidyverse:5358395134867368' }"

    input:
    tuple val(meta), path(idxstats), path(depth)
    path(references)

    output:
    tuple val(meta), path("*.csv"), path("*major.fa"), emit: major_mapping, optional: true
    tuple val(meta), path("*.csv"), path("*minor.fa"), emit: minor_mapping, optional: true
    tuple val(meta), path("*.csv"),                    emit: csv,           optional: true
    path "versions.yml",                               emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    summarize_mapping_to_all_references.R \\
        ${idxstats} \\
        ${depth} \\
        ${prefix} \\
        ${references} \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
      r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
      tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
      seqinr: \$(Rscript -e "library(seqinr); cat(as.character(packageVersion('seqinr')))")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Safe echo (interpolated by Groovy)
    echo "${args}"

    # Create outputs matching the declared emit patterns
    printf "reference,mapped_reads,unmapped_reads,depth_mean\n" > ${prefix}.mapping_summary.csv
    printf "3a_D17763,8079,1,94.3\n" >> ${prefix}.mapping_summary.csv
    printf "4k_EU392173,40,0,5.3\n" >> ${prefix}.mapping_summary.csv

    # Optional FASTA outputs (touch to create empty files)
    : > ${prefix}.major.fa
    : > ${prefix}.minor.fa

    # Stable versions file
    cat > versions.yml <<'YAML'
    "${task.process}":
      r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
      tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
      seqinr: \$(Rscript -e "library(seqinr); cat(as.character(packageVersion('seqinr')))")
    END_VERSIONS
    """
}
