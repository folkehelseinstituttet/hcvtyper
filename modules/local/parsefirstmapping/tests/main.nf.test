nextflow_process {

  name "Test Process PARSEFIRSTMAPPING"
  script "../main.nf"
  process "PARSEFIRSTMAPPING"

  tag "modules"
  tag "parsefirstmapping"

  // Real run with idxstats and depth files
  test("parsefirstmapping: idxstats + depth + references") {
    options "--profile docker"
    when {
      process {
        """
        // Input signature:
        // tuple val(meta), path(idxstats), path(depth)
        // path(references)

        input[0] = [
          [ id: 'Test_2' ],
          file("${projectDir}/modules/local/parsefirstmapping/tests/Test_2.firstmapping.withdup.idxstats", checkIfExists: true),
          file("${projectDir}/modules/local/parsefirstmapping/tests/Test_2.tsv", checkIfExists: true)
        ]
        input[1] = file("${projectDir}/modules/local/parsefirstmapping/tests/references.fa", checkIfExists: true)
        """
      }
    }
    then {
      assertAll(
        { assert process.success },
        { assert snapshot([
            csv        : process.out.csv,
            major_fasta: process.out.major_fasta,
            minor_fasta: process.out.minor_fasta,
            versions   : process.out.versions
          ]).match() }
      )
    }
  }

  // Fast stub run
  test("parsefirstmapping: stub") {
    options "-stub-run"
    when {
      process {
        """
        input[0] = [
          [ id: 'Test_2' ],
          file("${projectDir}/modules/local/parsefirstmapping/tests/Test_2.firstmapping.withdup.idxstats", checkIfExists: true),
          file("${projectDir}/modules/local/parsefirstmapping/tests/Test_2.tsv", checkIfExists: true)
        ]
        input[1] = file("${projectDir}/modules/local/parsefirstmapping/tests/references.fa", checkIfExists: true)
        """
      }
    }
    then {
      assertAll(
        { assert process.success },
        { assert snapshot([
            csv        : process.out.csv,
            major_fasta: process.out.major_fasta,
            minor_fasta: process.out.minor_fasta,
            versions   : process.out.versions
          ]).match() }
      )
    }
  }
}
