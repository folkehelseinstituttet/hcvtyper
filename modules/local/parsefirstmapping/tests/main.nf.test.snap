{
    "parsefirstmapping: idxstats + depth + references": {
        "content": [
            {
                "csv": [
                    [
                        {
                            "id": "Test_2"
                        },
                        "Test_2.parsefirstmapping.csv:md5,157308ac0d693c416c525a07ef6da263"
                    ]
                ],
                "major_fasta": [
                    [
                        {
                            "id": "Test_2"
                        },
                        "Test_2.3a_D17763_major.fa:md5,20ca3a85c0fff19e410596de95a7c45f"
                    ]
                ],
                "minor_fasta": [
                    [
                        {
                            "id": "Test_2"
                        },
                        "Test_2.4k_EU392173_minor.fa:md5,379bba7235f582f5ee3d83bc4fa9a4ca"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9daace31d09a7f33c621aa047da2bf44"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-08T11:48:35.302750481"
    },
    "parsefirstmapping: stub": {
        "content": [
            {
                "csv": [
                    [
                        {
                            "id": "Test_2"
                        },
                        "Test_2.mapping_summary.csv:md5,d4a863b19ab6607a0b2e2b2f5e6eb908"
                    ]
                ],
                "major_fasta": [
                    [
                        {
                            "id": "Test_2"
                        },
                        "Test_2.major.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "minor_fasta": [
                    [
                        {
                            "id": "Test_2"
                        },
                        "Test_2.minor.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,095949397f5779b287fd74d3aa01f318"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-08T11:48:43.196257864"
    }
}